File: clients/python/oms_client.py
#!/usr/bin/env python3
"""
OMS Client - Python SDK for Multi-Exchange OMS
"""

import grpc
import time
import argparse
from datetime import datetime
from typing import List, Optional, Dict, Any
import sys
import os

# Add proto path
sys.path.append(os.path.join(os.path.dirname(__file__), '../../proto'))
import oms_pb2
import oms_pb2_grpc


class OMSClient:
    """Python client for OMS gRPC server"""
    
    def __init__(self, server_address: str = "localhost:50051"):
        """Initialize OMS client
        
        Args:
            server_address: gRPC server address
        """
        self.channel = grpc.insecure_channel(server_address)
        self.stub = oms_pb2_grpc.OrderServiceStub(self.channel)
    
    def place_order(self, 
                   symbol: str,
                   side: str,
                   quantity: float,
                   order_type: str = "LIMIT",
                   price: float = 0,
                   exchange: str = "binance",
                   market: str = "spot",
                   account_id: str = "main") -> Dict[str, Any]:
        """Place a new order
        
        Args:
            symbol: Trading symbol (e.g., BTCUSDT)
            side: Order side (BUY or SELL)
            quantity: Order quantity
            order_type: Order type (LIMIT or MARKET)
            price: Order price (required for LIMIT orders)
            exchange: Exchange name
            market: Market type (spot or futures)
            account_id: Account ID
            
        Returns:
            Order response dictionary
        """
        request = oms_pb2.PlaceOrderRequest(
            symbol=symbol,
            side=side,
            order_type=order_type,
            quantity=quantity,
            price=price,
            exchange=exchange,
            market=market,
            account_id=account_id
        )
        
        response = self.stub.PlaceOrder(request)
        return {
            "order_id": response.order_id,
            "exchange_order_id": response.exchange_order_id,
            "status": response.status,
            "created_at": datetime.fromtimestamp(response.created_at)
        }
    
    def cancel_order(self, order_id: str) -> Dict[str, Any]:
        """Cancel an existing order
        
        Args:
            order_id: Order ID to cancel
            
        Returns:
            Cancel response dictionary
        """
        request = oms_pb2.CancelOrderRequest(order_id=order_id)
        response = self.stub.CancelOrder(request)
        
        return {
            "order_id": response.order_id,
            "status": response.status,
            "cancelled_at": datetime.fromtimestamp(response.cancelled_at)
        }
    
    def get_order(self, order_id: str) -> Dict[str, Any]:
        """Get order details
        
        Args:
            order_id: Order ID
            
        Returns:
            Order details dictionary
        """
        request = oms_pb2.GetOrderRequest(order_id=order_id)
        response = self.stub.GetOrder(request)
        
        return self._order_to_dict(response.order)
    
    def list_orders(self, status: str = "", symbol: str = "") -> List[Dict[str, Any]]:
        """List orders with optional filters
        
        Args:
            status: Filter by status (OPEN, FILLED, CANCELLED)
            symbol: Filter by symbol
            
        Returns:
            List of order dictionaries
        """
        request = oms_pb2.ListOrdersRequest(status=status, symbol=symbol)
        response = self.stub.ListOrders(request)
        
        return [self._order_to_dict(order) for order in response.orders]
    
    def get_balance(self, 
                   exchange: str = "binance",
                   market: str = "spot",
                   account_id: str = "main") -> Dict[str, Dict[str, float]]:
        """Get account balance
        
        Args:
            exchange: Exchange name
            market: Market type
            account_id: Account ID
            
        Returns:
            Balance dictionary {asset: {free, locked, total}}
        """
        request = oms_pb2.GetBalanceRequest(
            exchange=exchange,
            market=market,
            account_id=account_id
        )
        response = self.stub.GetBalance(request)
        
        balances = {}
        for balance in response.balances:
            balances[balance.asset] = {
                "free": balance.free,
                "locked": balance.locked,
                "total": balance.free + balance.locked
            }
        
        return balances
    
    def get_positions(self,
                     exchange: str = "binance",
                     account_id: str = "main") -> List[Dict[str, Any]]:
        """Get open positions (futures)
        
        Args:
            exchange: Exchange name
            account_id: Account ID
            
        Returns:
            List of position dictionaries
        """
        request = oms_pb2.GetPositionsRequest(
            exchange=exchange,
            account_id=account_id
        )
        response = self.stub.GetPositions(request)
        
        positions = []
        for pos in response.positions:
            positions.append({
                "symbol": pos.symbol,
                "side": pos.side,
                "size": pos.size,
                "entry_price": pos.entry_price,
                "mark_price": pos.mark_price,
                "unrealized_pnl": pos.unrealized_pnl,
                "pnl_percentage": pos.pnl_percentage,
                "leverage": pos.leverage,
                "margin": pos.margin
            })
        
        return positions
    
    def stream_prices(self, symbols: List[str], callback):
        """Stream real-time prices
        
        Args:
            symbols: List of symbols to stream
            callback: Callback function(price_update)
        """
        request = oms_pb2.StreamPricesRequest(symbols=symbols)
        
        for response in self.stub.StreamPrices(request):
            callback({
                "exchange": response.exchange,
                "symbol": response.symbol,
                "bid_price": response.bid_price,
                "bid_quantity": response.bid_quantity,
                "ask_price": response.ask_price,
                "ask_quantity": response.ask_quantity,
                "last_price": response.last_price,
                "timestamp": datetime.fromtimestamp(response.timestamp)
            })
    
    def stream_orders(self, callback):
        """Stream order updates
        
        Args:
            callback: Callback function(order_update)
        """
        request = oms_pb2.StreamOrdersRequest()
        
        for response in self.stub.StreamOrders(request):
            callback(self._order_to_dict(response.order))
    
    def _order_to_dict(self, order) -> Dict[str, Any]:
        """Convert protobuf order to dictionary"""
        return {
            "order_id": order.order_id,
            "exchange_order_id": order.exchange_order_id,
            "symbol": order.symbol,
            "side": order.side,
            "order_type": order.order_type,
            "quantity": order.quantity,
            "price": order.price,
            "filled_quantity": order.filled_quantity,
            "status": order.status,
            "exchange": order.exchange,
            "market": order.market,
            "account_id": order.account_id,
            "created_at": datetime.fromtimestamp(order.created_at),
            "updated_at": datetime.fromtimestamp(order.updated_at) if order.updated_at else None
        }
    
    def close(self):
        """Close the gRPC channel"""
        self.channel.close()


def main():
    """Example usage of OMS client"""
    parser = argparse.ArgumentParser(description="OMS Python Client")
    parser.add_argument("--server", default="localhost:50051", help="Server address")
    
    subparsers = parser.add_subparsers(dest="command", help="Commands")
    
    # Place order command
    place_parser = subparsers.add_parser("place", help="Place an order")
    place_parser.add_argument("--symbol", required=True, help="Trading symbol")
    place_parser.add_argument("--side", required=True, choices=["BUY", "SELL"], help="Order side")
    place_parser.add_argument("--quantity", type=float, required=True, help="Order quantity")
    place_parser.add_argument("--type", default="LIMIT", choices=["LIMIT", "MARKET"], help="Order type")
    place_parser.add_argument("--price", type=float, default=0, help="Order price")
    place_parser.add_argument("--exchange", default="binance", help="Exchange")
    place_parser.add_argument("--market", default="spot", choices=["spot", "futures"], help="Market type")
    place_parser.add_argument("--account", default="main", help="Account ID")
    
    # Cancel order command
    cancel_parser = subparsers.add_parser("cancel", help="Cancel an order")
    cancel_parser.add_argument("--id", required=True, help="Order ID")
    
    # Get order command
    get_parser = subparsers.add_parser("get", help="Get order details")
    get_parser.add_argument("--id", required=True, help="Order ID")
    
    # List orders command
    list_parser = subparsers.add_parser("list", help="List orders")
    list_parser.add_argument("--status", default="", help="Filter by status")
    list_parser.add_argument("--symbol", default="", help="Filter by symbol")
    
    # Balance command
    balance_parser = subparsers.add_parser("balance", help="Get balance")
    balance_parser.add_argument("--exchange", default="binance", help="Exchange")
    balance_parser.add_argument("--market", default="spot", help="Market type")
    balance_parser.add_argument("--account", default="main", help="Account ID")
    
    # Stream prices command
    stream_parser = subparsers.add_parser("stream", help="Stream prices")
    stream_parser.add_argument("--symbols", nargs="+", default=["BTCUSDT", "ETHUSDT"], help="Symbols to stream")
    
    args = parser.parse_args()
    
    if not args.command:
        parser.print_help()
        return
    
    # Create client
    client = OMSClient(args.server)
    
    try:
        if args.command == "place":
            result = client.place_order(
                symbol=args.symbol,
                side=args.side,
                quantity=args.quantity,
                order_type=args.type,
                price=args.price,
                exchange=args.exchange,
                market=args.market,
                account_id=args.account
            )
            print(f"Order placed successfully!")
            print(f"Order ID: {result['order_id']}")
            print(f"Status: {result['status']}")
            
        elif args.command == "cancel":
            result = client.cancel_order(args.id)
            print(f"Order cancelled successfully!")
            print(f"Order ID: {result['order_id']}")
            print(f"Status: {result['status']}")
            
        elif args.command == "get":
            order = client.get_order(args.id)
            print(f"Order Details:")
            for key, value in order.items():
                print(f"  {key}: {value}")
                
        elif args.command == "list":
            orders = client.list_orders(status=args.status, symbol=args.symbol)
            print(f"Found {len(orders)} orders:")
            for order in orders:
                print(f"\nOrder {order['order_id']}:")
                print(f"  Symbol: {order['symbol']} | Side: {order['side']} | Type: {order['order_type']}")
                print(f"  Quantity: {order['quantity']} | Filled: {order['filled_quantity']}")
                print(f"  Status: {order['status']}")
                
        elif args.command == "balance":
            balances = client.get_balance(
                exchange=args.exchange,
                market=args.market,
                account_id=args.account
            )
            print(f"Balance for {args.exchange} {args.market} (Account: {args.account}):")
            for asset, balance in balances.items():
                if balance['total'] > 0:
                    print(f"  {asset}: Free={balance['free']:.8f}, Locked={balance['locked']:.8f}, Total={balance['total']:.8f}")
                    
        elif args.command == "stream":
            print(f"Streaming prices for {args.symbols}... (Press Ctrl+C to stop)")
            
            def price_callback(update):
                print(f"[{update['timestamp'].strftime('%H:%M:%S')}] {update['exchange']} {update['symbol']}: "
                      f"Bid=${update['bid_price']:.2f} | Ask=${update['ask_price']:.2f} | Last=${update['last_price']:.2f}")
            
            client.stream_prices(args.symbols, price_callback)
            
    except grpc.RpcError as e:
        print(f"Error: {e.details()}")
    except KeyboardInterrupt:
        print("\nStopped by user")
    finally:
        client.close()


if __name__ == "__main__":
    main()

File: clients/javascript/oms-client.js
/**
 * OMS Client - JavaScript/Node.js SDK for Multi-Exchange OMS
 */

const grpc = require('@grpc/grpc-js');
const protoLoader = require('@grpc/proto-loader');
const path = require('path');

// Load proto file
const PROTO_PATH = path.join(__dirname, '../../proto/oms.proto');
const packageDefinition = protoLoader.loadSync(PROTO_PATH, {
  keepCase: true,
  longs: String,
  enums: String,
  defaults: true,
  oneofs: true
});

const omsProto = grpc.loadPackageDefinition(packageDefinition).oms;

class OMSClient {
  /**
   * Initialize OMS client
   * @param {string} serverAddress - gRPC server address (default: localhost:50051)
   */
  constructor(serverAddress = 'localhost:50051') {
    this.client = new omsProto.OrderService(
      serverAddress,
      grpc.credentials.createInsecure()
    );
  }

  /**
   * Place a new order
   * @param {Object} params - Order parameters
   * @returns {Promise<Object>} Order response
   */
  placeOrder(params) {
    return new Promise((resolve, reject) => {
      const request = {
        symbol: params.symbol,
        side: params.side,
        order_type: params.orderType || 'LIMIT',
        quantity: params.quantity,
        price: params.price || 0,
        exchange: params.exchange || 'binance',
        market: params.market || 'spot',
        account_id: params.accountId || 'main'
      };

      this.client.PlaceOrder(request, (err, response) => {
        if (err) {
          reject(err);
        } else {
          resolve({
            orderId: response.order_id,
            exchangeOrderId: response.exchange_order_id,
            status: response.status,
            createdAt: new Date(parseInt(response.created_at) * 1000)
          });
        }
      });
    });
  }

  /**
   * Cancel an existing order
   * @param {string} orderId - Order ID to cancel
   * @returns {Promise<Object>} Cancel response
   */
  cancelOrder(orderId) {
    return new Promise((resolve, reject) => {
      const request = { order_id: orderId };

      this.client.CancelOrder(request, (err, response) => {
        if (err) {
          reject(err);
        } else {
          resolve({
            orderId: response.order_id,
            status: response.status,
            cancelledAt: new Date(parseInt(response.cancelled_at) * 1000)
          });
        }
      });
    });
  }

  /**
   * Get order details
   * @param {string} orderId - Order ID
   * @returns {Promise<Object>} Order details
   */
  getOrder(orderId) {
    return new Promise((resolve, reject) => {
      const request = { order_id: orderId };

      this.client.GetOrder(request, (err, response) => {
        if (err) {
          reject(err);
        } else {
          resolve(this._formatOrder(response.order));
        }
      });
    });
  }

  /**
   * List orders with optional filters
   * @param {Object} filters - Filter options
   * @returns {Promise<Array>} List of orders
   */
  listOrders(filters = {}) {
    return new Promise((resolve, reject) => {
      const request = {
        status: filters.status || '',
        symbol: filters.symbol || ''
      };

      this.client.ListOrders(request, (err, response) => {
        if (err) {
          reject(err);
        } else {
          resolve(response.orders.map(order => this._formatOrder(order)));
        }
      });
    });
  }

  /**
   * Get account balance
   * @param {Object} params - Balance parameters
   * @returns {Promise<Object>} Balance object
   */
  getBalance(params = {}) {
    return new Promise((resolve, reject) => {
      const request = {
        exchange: params.exchange || 'binance',
        market: params.market || 'spot',
        account_id: params.accountId || 'main'
      };

      this.client.GetBalance(request, (err, response) => {
        if (err) {
          reject(err);
        } else {
          const balances = {};
          response.balances.forEach(balance => {
            balances[balance.asset] = {
              free: parseFloat(balance.free),
              locked: parseFloat(balance.locked),
              total: parseFloat(balance.free) + parseFloat(balance.locked)
            };
          });
          resolve(balances);
        }
      });
    });
  }

  /**
   * Get open positions (futures)
   * @param {Object} params - Position parameters
   * @returns {Promise<Array>} List of positions
   */
  getPositions(params = {}) {
    return new Promise((resolve, reject) => {
      const request = {
        exchange: params.exchange || 'binance',
        account_id: params.accountId || 'main'
      };

      this.client.GetPositions(request, (err, response) => {
        if (err) {
          reject(err);
        } else {
          const positions = response.positions.map(pos => ({
            symbol: pos.symbol,
            side: pos.side,
            size: parseFloat(pos.size),
            entryPrice: parseFloat(pos.entry_price),
            markPrice: parseFloat(pos.mark_price),
            unrealizedPnl: parseFloat(pos.unrealized_pnl),
            pnlPercentage: parseFloat(pos.pnl_percentage),
            leverage: parseInt(pos.leverage),
            margin: parseFloat(pos.margin)
          }));
          resolve(positions);
        }
      });
    });
  }

  /**
   * Stream real-time prices
   * @param {Array<string>} symbols - Symbols to stream
   * @param {Function} callback - Callback for price updates
   * @returns {Object} Stream object
   */
  streamPrices(symbols, callback) {
    const request = { symbols };
    const stream = this.client.StreamPrices(request);

    stream.on('data', (response) => {
      callback({
        exchange: response.exchange,
        symbol: response.symbol,
        bidPrice: parseFloat(response.bid_price),
        bidQuantity: parseFloat(response.bid_quantity),
        askPrice: parseFloat(response.ask_price),
        askQuantity: parseFloat(response.ask_quantity),
        lastPrice: parseFloat(response.last_price),
        timestamp: new Date(parseInt(response.timestamp) * 1000)
      });
    });

    stream.on('error', (err) => {
      console.error('Stream error:', err);
    });

    stream.on('end', () => {
      console.log('Stream ended');
    });

    return stream;
  }

  /**
   * Stream order updates
   * @param {Function} callback - Callback for order updates
   * @returns {Object} Stream object
   */
  streamOrders(callback) {
    const request = {};
    const stream = this.client.StreamOrders(request);

    stream.on('data', (response) => {
      callback(this._formatOrder(response.order));
    });

    stream.on('error', (err) => {
      console.error('Stream error:', err);
    });

    stream.on('end', () => {
      console.log('Stream ended');
    });

    return stream;
  }

  /**
   * Format order object
   * @private
   */
  _formatOrder(order) {
    return {
      orderId: order.order_id,
      exchangeOrderId: order.exchange_order_id,
      symbol: order.symbol,
      side: order.side,
      orderType: order.order_type,
      quantity: parseFloat(order.quantity),
      price: parseFloat(order.price),
      filledQuantity: parseFloat(order.filled_quantity),
      status: order.status,
      exchange: order.exchange,
      market: order.market,
      accountId: order.account_id,
      createdAt: new Date(parseInt(order.created_at) * 1000),
      updatedAt: order.updated_at ? new Date(parseInt(order.updated_at) * 1000) : null
    };
  }
}

// Example usage
async function example() {
  const client = new OMSClient('localhost:50051');

  try {
    // Place a limit order
    console.log('Placing order...');
    const orderResult = await client.placeOrder({
      symbol: 'BTCUSDT',
      side: 'BUY',
      quantity: 0.001,
      price: 115000,
      orderType: 'LIMIT'
    });
    console.log('Order placed:', orderResult);

    // Get order details
    const order = await client.getOrder(orderResult.orderId);
    console.log('Order details:', order);

    // Get balance
    const balance = await client.getBalance();
    console.log('Balance:', balance);

    // Stream prices
    console.log('Streaming prices...');
    const priceStream = client.streamPrices(['BTCUSDT', 'ETHUSDT'], (update) => {
      console.log(`Price update: ${update.symbol} - Bid: $${update.bidPrice.toFixed(2)}, Ask: $${update.askPrice.toFixed(2)}`);
    });

    // Stop streaming after 10 seconds
    setTimeout(() => {
      priceStream.cancel();
      console.log('Stopped streaming');
    }, 10000);

  } catch (error) {
    console.error('Error:', error);
  }
}

// Export client
module.exports = OMSClient;

// Run example if called directly
if (require.main === module) {
  example().catch(console.error);
}